ocaml/xapi/xapi_session.ml

```diff
@@ -442,53 +441,64 @@ let revalidate_external_session ~__context ~session =
             (trackid session)
         in
         debug "%s" msg ;
-        destroy_db_session ~__context ~self:session
+        destroy_db_session ~__context ~self:session ;
+        acc
       ) else
         (* otherwise, we try to revalidate it against the external authentication service *)
         let session_lifespan = 60.0 *. 30.0 in
         (* allowed session lifespan = 30 minutes *)
         let random_lifespan = Random.float 60.0 *. 10.0 in
 
-        (* extra random (up to 10min) lifespan to spread access to external directory *)
-
         (* 2. has the external session expired/does it need revalidation? *)
         let session_last_validation_time =
           Date.to_unix_time
             (Db.Session.get_validation_time ~__context ~self:session)
         in
         let now = Date.now () in
-        let session_needs_revalidation =
+        let session_timeout =
           Date.to_unix_time now
           > session_last_validation_time +. session_lifespan +. random_lifespan
         in
-        if session_needs_revalidation then (
+
+        (* extra random (up to 10min) lifespan to spread access to external directory *)
+        let authenticated_user_sid =
+          Db.Session.get_auth_user_sid ~__context ~self:session
+        in
+        let user_validated =
+          (* acc is [(sid, check_result)] , true for check pass, false for check failed *)
+          match List.assoc_opt authenticated_user_sid acc with
+          | None ->
+              false
+          | Some v ->
+              if v = false && session_timeout then (
+                debug
+                  "Destory session %s as previous check for user %s not pass"
+                  (trackid session) authenticated_user_sid ;
+                destroy_db_session ~__context ~self:session
+              ) ;
+              debug "Skip check session %s as previous check for user %s exists"
+                (trackid session) authenticated_user_sid ;
+              true
+        in
+
+        if session_timeout && not user_validated then (
```

Comment:
Could this be the following? The pattern matching and if condition in `user_validated` seem not easy to understand.

```
let validate_with_memo acc f =
  match List.assoc_opt authenticated_user_sid acc with
  | None ->
      f acc
  | Some false ->
      destroy_db_session ~__context ~self:session ;
      acc
  | Some true ->
      acc
in

if session_timeout then (
  validate_with_memo acc @@
  ...
```

Reply:
No. It will not add more git diffs.